src/hp/netcheck.rs

```diff
@@ -1106,85 +1065,298 @@ impl ProbeReport {
             .await
-            .context("failed to create ReportState")?;
-        let (in_flight_s, mut in_flight_r) = sync::mpsc::channel(8);
-        let mut running = {
-            let port_mapper = self.port_mapper.clone();
-            let skip_external = self.skip_external_network;
-            let resolver = self.dns_resolver.clone(); // not a cheap clone
-            Box::pin(time::timeout(OVERALL_PROBE_TIMEOUT, async move {
-                report_state
-                    .run(in_flight_s, dm, port_mapper, skip_external, &resolver)
+        {
+            Ok(()) => {
+                self.current_check_run = Some(response_tx);
+            }
+            Err(err) => {
+                response_tx.send(Err(err)).ok();
+            }
+        }
+    }
+
+    /// Spawns a task running a [`ReportState`] run.
+    ///
+    /// When the run is completed the task sends the result back to this actor.
+    async fn start_report_run(
+        &mut self,
+        derp_map: DerpMap,
+        stun_sock_v4: Option<Arc<UdpSocket>>,
+        stun_sock_v6: Option<Arc<UdpSocket>>,
+    ) -> Result<()> {
+        // TODO: These spawns should not be fatal on failure, see issue #1025
+        let cancel_token = CancellationToken::new();
+        let stun_sock_v4 = match stun_sock_v4 {
+            Some(stun_sock_v4) => stun_sock_v4,
+            None => {
+                let addr = SocketAddr::from((Ipv4Addr::UNSPECIFIED, 0));
+                let sock = UdpSocket::bind(addr)
                     .await
-            }))
+                    .context("netcheck: failed to bind udp 0.0.0.0:0")?;
+                let sock = Arc::new(sock);
+                spawn_udp_listener(sock.clone(), self.addr(), cancel_token.clone());
+                sock
+            }
+        };
+        let stun_sock_v6 = match stun_sock_v6 {
+            Some(stun_sock_v6) => stun_sock_v6,
+            None => {
+                let addr = SocketAddr::from((Ipv6Addr::UNSPECIFIED, 0));
+                let sock = UdpSocket::bind(addr)
+                    .await
+                    .context("netcheck: failed to bind udp6 [::]:0")?;
+                let sock = Arc::new(sock);
+                spawn_udp_listener(sock.clone(), self.addr(), cancel_token.clone());
+                sock
+            }
         };
-        let mut in_flight = HashMap::new();
 
-        loop {
-            tokio::select! {
-                Some(inf) = in_flight_r.recv() => {
-                    in_flight.insert(inf.tx, inf);
+        let report_state = self
+            .create_report_state(&derp_map, stun_sock_v4, stun_sock_v6)
+            .await
+            .context("failed to create ReportState")?;
+        let port_mapper = self.port_mapper.clone();
+        let skip_external = self.skip_external_network;
+        let resolver = self.dns_resolver.clone(); // not a cheap clone, maybe Arc it
+        let addr = self.addr();
+        tokio::spawn(async move {
```

Comment:
Would be nice to track this task, and make sure it gets aborted on drop as well

Reply:
Yeah, I was wondering about that as well.  But kind of wanted to defer that to a followup PR.  Created https://github.com/n0-computer/iroh/issues/1037 for this as there's a few options to consider.